modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+        fail_with(Failure::UnexpectedReply, result_message)
+      end
+
+      if returned_entries.blank?
+        print_error("No results found for #{filter}.")
+
+        nil
+      else
+
+        returned_entries
+      end
+    end
+  rescue Rex::ConnectionTimeout
+    fail_with(Failure::Unreachable, "Couldn't reach #{datastore['RHOST']}!")
+  rescue Net::LDAP::Error => e
+    fail_with(Failure::UnexpectedReply, "Could not query #{datastore['RHOST']}! Error was: #{e.message}")
+  end
+
+  def convert_sids_to_human_readable_name(sids_array)
+    output = []
+    for sid in sids_array
+      raw_filter = "(objectSID=#{sid})"
+      attributes = ['dn', 'description']
+      sid_entry = query_ldap_server(raw_filter, attributes)
+      fail_with(Failure::UnexpectedReply, 'Could not find any matching SID!') if sid_entry.blank?
+      if sid_entry[0][:description]
+        output << [sid, sid_entry[0][:dn][0], sid_entry[0][:description][0]]
+      else
+        output << [sid, sid_entry[0][:dn][0], nil]
+      end
+    end
+
+    result = ''
+    output.each do |sid_string, dn, description|
+      if description
+        result += "#{sid_string} aka #{dn} (#{description}) | "
+      else
+        result += "#{sid_string} aka #{dn} | "
+      end
+    end
+
+    result.gsub(%r{ \| $}, '') # Strip out the last trailing " \ " from the output.
+  end
+
+  def find_esc1_vuln_cert_templates
+    esc1_raw_filter = '(&'\
+      '(objectclass=pkicertificatetemplate)'\
+      '(!(mspki-enrollment-flag:1.2.840.113556.1.4.804:=2))'\
+      '(|(mspki-ra-signature=0)(!(mspki-ra-signature=*)))'\
+      '(|'\
+        '(pkiextendedkeyusage=1.3.6.1.4.1.311.20.2.2)'\
+        '(pkiextendedkeyusage=1.3.6.1.5.5.7.3.2)'\
+        '(pkiextendedkeyusage=1.3.6.1.5.2.3.4)'\
+        '(pkiextendedkeyusage=2.5.29.37.0)'\
+        '(!(pkiextendedkeyusage=*))'\
+      ')'\
+      '(mspki-certificate-name-flag:1.2.840.113556.1.4.804:=1)'\
+    ')'
+    attributes = ['cn', 'description', 'ntSecurityDescriptor']
+    base_prefix = 'CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration'
+    esc1_entries = query_ldap_server(esc1_raw_filter, attributes, base_prefix: base_prefix)
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC1 vulnerable certificate templates!') if esc1_entries.blank?
+
+    # Grab a list of certificates that contain vulnerable settings.
+    # Also print out the list of SIDs that can enroll in that server.
+
+    esc1_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
```

Comment:
`flag_allowed_to_enroll` will only be set if `security_descriptor.dacl`. The value will be kept through the next loop iteration if `security_descriptor.dacl` is nil, which would lead to false positives. You should always initialize `flag_allowed_to_enroll` to fall on each loop iteration. That will ensure that the only way it's `true` is if `security_descriptor.dacl` is truthy *and* `parse_dacl_or_sacl` returns true for it's first value.

```suggestion
      flag_allowed_to_enroll = false
      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
      if flag_allowed_to_enroll
```

If that looks a little odd, you could alternatively use:

```suggestion
      if security_descriptor.dacl
        flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) 
        if flag_allowed_to_enroll
```

Reply:
Fixed this up with https://github.com/rapid7/metasploit-framework/commit/fd33cdc58e47576823a874c5f17d81b3e5660f9b which uses the new function `query_ldap_server_certificates` you recommended to reduce code reuse and also added another bit of code to fix the `find_enrollable_vuln_certificate_templates` function to ensure that the flag is reset on every iteration.